docs/networkfenceclass.md

```diff
@@ -0,0 +1,61 @@
+# NetworkFenceClass
+
+NetworkFence is a cluster-scoped custom resource that allows Kubernetes to invoke "GetFenceClients" operation on a storage provider.
+
+The user needs to specify the csi provisioner name, parameters and the secret required to perform GetFenceClients operation.
+
+## Fence Operation
+
+```yaml
+apiVersion: csiaddons.openshift.io/v1alpha1
+kind: NetworkFenceClass
+metadata:
+  name: network-fence-class
+spec:
+  provisioner: example.driver
+  parameters:
+    key: value
+    csiaddons.openshift.io/networkfence-secret-name: secret-name
+    csiaddons.openshift.io/networkfence-secret-namespace: secret-namespace
+```
+
+- `provisioner`: specifies the name of storage provisioner.
+- `parameters`: specifies storage provider specific parameters.
+
+Resereved parameters:
+
+- `csiaddons.openshift.io/networkfence-secret-name`: specifies the name of the secret required for network fencing operation.
+- `csiaddons.openshift.io/networkfence-secret-namespace`: specifies the namespace in which the secret is located.
+
+Once the NetworkFenceClass is processed, the CSI Addons controller will call the GetFenceClients operation on the storage provider associated with the provisioner name that registered the `GET_CLIENTS_TO_FENCE` capability. The resulting data will then be stored in the CSIAddonsNode status.
+
+The NetworkFenceStatus object will contain the list of clients that need to be fenced.
+
+```yaml
+apiVersion: csiaddons.openshift.io/v1alpha1
+kind: CSIAddonsNode
+metadata:
+  annotations:
+    csiaddons.openshift.io/networkfenceclass-0: network-fence-class
```

Comment:
why the `-0` postfix?

Reply:
Is there really a need for such an ugly annotation? Can't whatever is going to fence not check the `CSIAddonsNode.spec.drivername` and match it with whatever NetworkFenceClasses there are? The creator of a NetworkFence CR will still need to pick a NetworkFenceClass for it, right?

Looping through such annotation is not really less work than looping through the NetworkFenceClasses and finding the right driver. Making sure the annotations are always correct seems more work than useful, which can have additional bugs with severe results (unable to fence).

Reply:
@nixpanic annotations are used as a way to trigger a reconcile. i am trying to avoid list operation and also there can be causes where the NFC classes are created/delete/recreate later after the csiaddosnnode registration is done, These are only of the way to triggering the Reconcile, The User will still need to look into the driverName and the NFC name present in the status field not on the annotation to get the right client Ip cidr to fence (This can be a documentation improvement i can do).  another option is to have something below but that can have length limitation in some worst cases scenarios.

```yaml
metadata:
  annotations:
    csiaddons.openshift.io/networkfenceclassname: '[{"name":"nfcName1"},{"name":"nfcName2"}]'
```

Reply:
I must be missing something, what is the reason a reconcile is needed? When a NetworkFenceClass has a matching drivername, it is expected that the CSIAddonsNode supports that class.

CSI Provisioners and NodePlugins also do not need to have annotations for the different StorageClasses, why is this different?

Reply:
The goal is to advertise the Ip's on the CSIAddonsNode and use the NFCClass to get the cluster details to get the client IP from. In PR description i explained the workflow. This is what we have discussed early (let me know if thats not the case).

Reply:
Okay, so what I was missing (or forgetting) is that the NetworkFenceClass-Controller is not the owner of the CSIAddonsNode CR. Therefor it should not update the status of the CSIAddonsNode with the IP-address(es) that were detected by the NetworkFenceClass trigger.

Instead, the NetworkFenceClass-Controller adds an annotation, so that the CSIAddonsNode-Controller is informed to call the `GetFenceClients()` CSI-Addons procedure of the CSI-driver and update the `CSIAddonsNode/status` after that.